### PR TITLE
Introduce JEP for External Fingerprint Storage

### PR DESCRIPTION
**Pull request by stellargo**
CC @oleg-nenashev @afalko @mikecirioli

### CODE REVIEW

File: jep/0000/README.adoc [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+This JEP builds such a reference implementation for Redis.
+
+== Specification
+
+=== Top-level requirements
+
+* Fingerprint Storage has a pluggable architecture
+* Reference implementations:
+
+** Local XML-based file system storage (current implementation)
```
Comment (oleg-nenashev) [inline jep/0000/README.adoc:78]:
```suggestion
** Local XML-based file system storage (current implementation) - default implementation in the Jenkins core
```

File: jep/0000/README.adoc [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+Testing is relatively straightforward using JUnit, JTH, testcontainers, etc.
+
+== Prototype Implementation
+
+link:https://github.com/jenkinsci/jenkins/pull/4731[PR in Jenkins Core for Externalizing Fingerprint Storage]
+
+link:https://github.com/jenkinsci/redis-fingerprint-storage-plugin[Reference Implementation]
+
+== References
+
```
Comment (oleg-nenashev) [inline jep/0000/README.adoc:289]: Convert to the list?

File: jep/0000/README.adoc [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+This proposal does not define strong security requirements for external fingerprint storage implementations.
+
+== Infrastructure Requirements
+
+There are no new infrastructure requirements related to this proposal.
+
+== Testing
+
+Testing is relatively straightforward using JUnit, JTH, testcontainers, etc.
```
Comment (oleg-nenashev) [inline jep/0000/README.adoc:280]: Famous last words, but OK. It would be great to expand it later so that a reader could understand the scope

File: jep/0000/README.adoc [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+
+== Prototype Implementation
+
+link:https://github.com/jenkinsci/jenkins/pull/4731[PR in Jenkins Core for Externalizing Fingerprint Storage]
+
+link:https://github.com/jenkinsci/redis-fingerprint-storage-plugin[Reference Implementation]
+
+== References
+
+link:https://docs.google.com/document/d/10f3IXTA6UMLUOFMTH_atQ3XlyWB3S7KGNCtTZmOUGdM/edit#[GSoC Proposal]
```
Comment (oleg-nenashev) [inline jep/0000/README.adoc:290]: Would it make sense to link the design doc instead?

File: jep/0000/README.adoc [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+
+However, the API includes functionality to retrieve fingerprints from other instances in the case when they are required for tracing purposes.
+This allows for tracing fingerprints across Jenkins instances.
+Implementation wise, this is achieved by plugins ensuring that Fingerprints have an associated Jenkins Instance ID.
+
+==== Local to External Migration
+
+When a plugin is configured and enabled, the fingerprints are transferred from the local storage to the external storage.
+
+This by followed by clearing the local database of fingerprints from Jenkins HOME ONLY if the writes are successful.
```
Comment (afalko) [inline jep/0000/README.adoc:115]: This is followed?

File: jep/0000/README.adoc [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+
+The solution should be generic enough in order to support common storage types. The following storage types are supported:
+
+* Local XML-based file system storage (current implementation)
+* Redis backed Fingerprint Storage Plugin
+
+The API does not force a particular ORM mapping for the storages.
+The storage plugins are passed on the `Fingerprint` objects, leaving it up to them to decide how to save the fingerprints.
+
+==== Instance Specificness of Fingerprints
```
Comment (afalko) [inline jep/0000/README.adoc:95]: I'd reword this header. Maybe: "Finger 
prints across instances"

File: jep/0000/README.adoc [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+This proposal offers the following solution:
+
+* When a build/job is deleted from Jenkins, a subsequent update is performed to the fingerprint to reflect this change
+* An async periodic work is run to check if any fingerprint has no jobs/builds associated with it and delete it. (But, it would ensure that if there is a deletion blocking `FingerprintFacet`, then the fingerprint would not be deleted.)
+
+The above solution is provided to the user as an option, and can be toggled because it may be the case that storing data (i.e. fingerprints) might be cheaper than to do consistent updates in the database.
+
+We do not propose to revamp the `FingerprintCleanupThread` to use the above logic at the moment.
+
+==== External to Local Migration (Nice 2 Have)
```
Comment (afalko) [inline jep/0000/README.adoc:132]: Change "nice 2 have"s to "out of scope for MVP" ... or at the very least, s/2/to/ ;)

Comment (afalko) [reply jep/0000/README.adoc:132]: I'd probably actually move these sections to the bottom so they don't distract from the important stuff.

File: jep/0000/README.adoc [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+
+== Motivation
+
+File fingerprinting is a way to track which version of a file is being used by a job/build, making dependency tracking easy.
+The fingerprint engine of Jenkins can track usages of artifacts, credentials, files, etc. within the system.
+Currently, it does this by maintaining a local XML-based database which leads to dependence on the physical disk of the Jenkins master.
+
+Allowing fingerprint storage to be moved to external storages decreases the dependence of Jenkins instances on the physical disk space and also allows for tracking the flow of fingerprints across instances of Jenkins connected to the same external storage.
+
+External storages:
```
Comment (afalko) [inline jep/0000/README.adoc:230]: "Advantages of using external storage drivers:

File: jep/0000/README.adoc (hunk 5 above)
Comment (stellargo) [reply jep/0000/README.adoc:115]: Hi @afalko, thanks for reviewing. No this is not followed yet, but according to JEP's guidelines, we need to frame it in the present tense as if it is already done [[source](https://github.com/jenkinsci/jep/tree/master/jep-template/0000)].

File: jep/0000/README.adoc (hunk 7 above)
Comment (stellargo) [reply jep/0000/README.adoc]: @afalko thanks! Incorporated "out of scope for MVP" :) 
I have these at the end of the Design Decisions section

**Comment by oleg-nenashev** [comment]
I will go ahead and merge it

File: jep/0000/README.adoc [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+
+==== NEW: `FileFingerprintStorage` class
+
+The current XML based local file storage is moved over to `FileFingerprintStorage`, which implements `FingerprintStorage`.
+
+This is the default `FingerprintStorage` provided, when no external pluggable storage plugin has been configured.
+
+==== `SaveableListener`
+
+As mentioned in this link:https://issues.jenkins-ci.org/browse/JENKINS-62543[JIRA issue], the current API of `SaveableListener` requires an `XMLFile` as an argument, which is not entirely practical for external storages.
```
Comment (jglick) [inline jep/0000/README.adoc:213]:
```suggestion
As mentioned in this link:https://issues.jenkins-ci.org/browse/JENKINS-62543[JIRA issue], the current API of `SaveableListener` requires an `XmlFile` as an argument, which is not entirely practical for external storages.
```

**Comment by oleg-nenashev** [comment]
JEP-225 ID is locked by #266. It does not really follow the JEP process, but I do not want to provoke additional merge conflicts. We are not required to do sequential numbering. I will merge one as JEP-226.

File: jep/226/README.adoc [hunk 10]
```diff
@@ -1,4 +1,4 @@
-= JEP-0000: External Fingerprint Storage
+= JEP-226: External Fingerprint Storage
```
Comment (jglick) [inline jep/226/README.adoc:1]: @oleg-nenashev will you update https://www.jenkins.io/sigs/cloud-native/pluggable-storage/ to match?

Comment (stellargo) [reply jep/226/README.adoc:1]: Thanks for pointing it out @jglick, Ill submit a PR to update it

Comment (oleg-nenashev) [reply jep/226/README.adoc:1]: Somebody will do it, if course.

Comment (stellargo) [reply jep/226/README.adoc:1]: https://github.com/jenkins-infra/jenkins.io/pull/3449
